File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/
import java.util.HashMap;

class Solution {
    public Node copyRandomList(Node head) {
        HashMap<Node,Node> old = new HashMap<Node, Node>();
        old.put(null, null);
        
        Node cur = head;
        while(cur != null){
            Node copy = new Node(cur.val);
            old.put(cur, copy);
            cur = cur.next;
        }
        
        cur = head;
        while(cur != null){
            Node copy = old.get(cur);
            copy.next = old.get(cur.next);
            copy.random = old.get(cur.random);
            old.put(cur, copy);
            cur = cur.next;
        }

        return old.get(head);
    }
}

File: 138-copy-list-with-random-pointer/copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

#include<unordered_map>

class Solution {
public:
    Node* copyRandomList(Node* head) {
        unordered_map<Node*,Node*> old;
        old[nullptr] = nullptr;
        
        Node* cur = head;
        while(cur){
            Node* copy = new Node(cur->val);
            old[cur] = copy;
            cur = cur->next;
        }
        
        cur = head;
        while(cur){
            old[cur]->next = old[cur->next];
            old[cur]->random = old[cur->random];
            cur = cur->next;
        }

        return old[head];
    }
};